### PR TITLE
bpo-35805: Add parser for Message-ID email header.

### PR DESCRIPTION
**Pull request by maxking**
This parser is based on the definition of Identification Fields from RFC 5322
Sec 3.6.4.

This should also prevent folding of Message-ID header using RFC 2047 encoded
words and hence fix [bpo-35805](https://bugs.python.org/issue35805).

https://bugs.python.org/issue35805

### CODE REVIEW

**Comment by maxking** [comment]
This only adds parser for `Message-ID` header for now. After this is reviewed/merged I can work on adding other "Identification Fields" from RFC 5322 using the same parser for msg_id.

File: Lib/email/_header_value_parser.py
Comment (eamanu) [inline Lib/email/_header_value_parser.py]: I think will be better remove this unused line.

And do the same with the other methods.

File: Lib/email/_header_value_parser.py
Comment (eamanu) [inline Lib/email/_header_value_parser.py]: What about have the docstring on one line?

```
def get_no_fold_literal(value):
    """no-fold-literal = "[" *dtext "]""""
````

?

File: Lib/email/_header_value_parser.py
Comment (warsaw) [inline Lib/email/_header_value_parser.py:2050]: Or should this append a defect?

**Review by warsaw — APPROVED** [review]
I realize sadly I no longer remember how this all works, but based on my recollection, this looks good to me.  I would be interested in @bitdancer 's opinion too.

File: Lib/email/_header_value_parser.py
Comment (warsaw) [reply Lib/email/_header_value_parser.py]: Agreed.

Comment (maxking) [reply Lib/email/_header_value_parser.py]: I just wanted to make it look consistent with the other classes above. All of them have an empty line.

File: Lib/email/_header_value_parser.py
Comment (maxking) [reply Lib/email/_header_value_parser.py:2050]: I was a bit confused on when to raise and when to append a defect.

I looked at `parse_addr_spec` to choose what behaviour should be, but I guess Address has more importance for routing than `Message-ID`.  I can change to to append a defect if that is better.

File: Lib/email/_header_value_parser.py
Comment (maxking) [reply Lib/email/_header_value_parser.py]: Do you think this would work?
```
"""no-fold-literal = "[" *dtext "]

"""
```
Again, just for consistency. Other methods in this module use the same format.

File: Lib/email/_header_value_parser.py
Comment (maxking) [reply Lib/email/_header_value_parser.py:2050]: Although, @bitdancer did mention to me that we don't raise exceptions when parsing and try to as best as we can. So maybe I should just add a defect. 

Will fix that.

Comment (maxking) [reply Lib/email/_header_value_parser.py:2050]: Fixed this by catching this exception in `parse_message_id` instead.

File: Lib/email/headerregistry.py
Comment (tirkarthi) [inline Lib/email/headerregistry.py:557]: Slightly offtopic to PR but during review I find `_default_header_map` is documented at https://docs.python.org/3/library/email.headerregistry.html#email.headerregistry.HeaderRegistry but is out of sync with the current default map. Should this be updated? If so it can be done as separate PR.

File: Lib/email/_header_value_parser.py
Comment (bitdancer) [reply Lib/email/_header_value_parser.py]: I'd be fine with deleting all those blank lines, but I put them in because at the time I thought that's what PEP 8 recommended, and some of the TokenList subclasses do have additional methods and in that case I think it looks better.

File: Lib/email/_header_value_parser.py
Comment (bitdancer) [reply Lib/email/_header_value_parser.py:2050]: Yes, that is the correct place to catch it :)  Basically the rule is the 'get' functions raise errors so that higher level parsing functions can catch them and try to parse one of the other possible token types that could be at that location, but 'parse' methods can't raise errors, they have to make the final decision as to how to handle bad data.  But I think you figured that part out.  When to raise rather than register a defect is a "decision of art", such as you made when deciding to return-with-defect for a missing id-right if there's a '>'.  I think you made the right decisions here, but you might need to revisit them when using get_msg_id to parse other headers.  The "art" is doing our best to do the "Postel Principle" dance here, and it is very much an art, not a science :)

File: Lib/email/headerregistry.py
Comment (bitdancer) [reply Lib/email/headerregistry.py:557]: Yes, it should be updated.  And you might as well do it in this PR, since that doc should be updated anyway to reflect this new feature :)

File: Lib/test/test_email/test_headerregistry.py
Comment (bitdancer) [inline Lib/test/test_email/test_headerregistry.py:1657]: Should have another test with a no-fold-literal in it.  And what happens if we get a header with non-ascii in the message-id?

**Review by bitdancer — DISMISSED** [review]
Overall this looks great.  Thanks!

**Comment by bitdancer** [comment]
@warsaw: I'm not sure *I* remember how it all works, and I wrote it.  As far as I can tell @maxking did a great job of figuring out the principles behind this rather baroque parser :)

For the record, this parser is essentially a "first draft" (or maybe a 1.5 draft, I forget), which I had hoped to come back to, extract what I learned from writing it, and write a version 2 that was more consistent and easier to understand (and most importantly would produce a parse tree that was easier to interrogate and manipulate).  But at this point it seems unlikely I'll ever manage to find time to do that.

File: Lib/email/_header_value_parser.py
Comment (maxking) [reply Lib/email/_header_value_parser.py:2050]: Thanks David for the explanation!

File: Lib/email/_header_value_parser.py
Comment (maxking) [reply Lib/email/_header_value_parser.py]: Done! Removed empty lines in all the classes that don't have a method.

File: Lib/email/headerregistry.py
Comment (maxking) [reply Lib/email/headerregistry.py:557]: Added `message-id` and all the missing headers too. Thanks @tirkarthi for pointing it out.

File: Lib/test/test_email/test_headerregistry.py
Comment (maxking) [reply Lib/test/test_email/test_headerregistry.py:1657]: Thanks for pointing it out!

As it turns out, the message-id not folding was completely by chance since all the tested subtokens weren't allowed to be folded and hence the header wasn't folding. As soon as I added a test with non-ascii characters, they started to fold.

I read through `_refold_parse_tree` to figure out if there is a good way to prevent folding of **subparts** when the part itself isn't allowed to be folded. See my new change for what I ended up doing. 

A side effect of which is that the Header does get wrapped if it exceeds the max_ilne_length but the msg-id token is kept intact (which will be needed when parsing `References` header). We also don't do any decoding of non-ascii parts. 

I did another implementation where I added `fold` method to `Message-ID` header to just return `str(header) + policy.linesep`, but that seemed wrong since there was possibility that we could fold at the boundary between `Message-ID: ` and `value`. Also, that probably won't work well for other Header types like `References` which would have multiple `msg-id` tokens.

What do you think about the current implementation? It feels a bit hacky to be honest, but I can't find a better way.

**Comment by maxking** [comment]
I have made the requested changes; please review again

File: Lib/test/test_email/test_headerregistry.py
Comment (maxking) [reply Lib/test/test_email/test_headerregistry.py:1657]: I am slightly confused with the actual RFC text, since it says:

   > Unlike addr-spec, this syntax only permits the dot-atom-text
   > form on the left-hand side of the "@" and does not have internal CFWS
   > anywhere in the message identifier.

But it also allows `obs-id-left`  and `obs-id-right` which is allowed to be folded using encoded words.  So do we allow folding of subparts or not?

Also not sure if leaving non-ascii text as is great idea.

Comment (bitdancer) [reply Lib/test/test_email/test_headerregistry.py:1657]: I think what you want is a fold method on MsgID that returns str(self) (with a comment that msg-ids may not be folded).  I unfortunately can't remember what the Right Thing is to do with non-ascii, so I'll probably have to take a deeper look at that.

**Comment by maxking** [comment]
@bitdancer I have made the changes you requested for the folding of msg-id tokens.

File: Lib/email/_header_value_parser.py
Comment (xnox) [inline Lib/email/_header_value_parser.py:2075]: This appears to bomb out when building hyper kitty, because token is referenced before assignment, when `try` block fails.

```
======================================================================
ERROR: test_long_message_id (hyperkitty.tests.lib.test_incoming.TestAddToList)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "./hyperkitty/tests/lib/test_incoming.py", line 295, in test_long_message_id
    msg["Message-ID"] = "X" * 260
  File "/usr/lib/python3.8/email/message.py", line 409, in __setitem__
    self._headers.append(self.policy.header_store_parse(name, val))
  File "/usr/lib/python3.8/email/policy.py", line 148, in header_store_parse
    return (name, self.header_factory(name, value))
  File "/usr/lib/python3.8/email/headerregistry.py", line 602, in __call__
    return self[name](name, value)
  File "/usr/lib/python3.8/email/headerregistry.py", line 197, in __new__
    cls.parse(value, kwds)
  File "/usr/lib/python3.8/email/headerregistry.py", line 530, in parse
    kwds['parse_tree'] = parse_tree = cls.value_parser(value)
  File "/usr/lib/python3.8/email/_header_value_parser.py", line 2116, in parse_message_id
    message_id.append(token)
UnboundLocalError: local variable 'token' referenced before assignment
```

Comment (surkova) [reply Lib/email/_header_value_parser.py:2075]: Yep, same for me. Also, accessing `value[0]` before checking if value is there…

Comment (xnox) [reply Lib/email/_header_value_parser.py:2075]: @surkova Thank you! I think this should be fixed up properly. Opened BPO https://bugs.python.org/issue38708

**Comment by odony** [comment]
> After this is reviewed/merged I can work on adding other "Identification Fields" from RFC 5322 using the same parser for msg_id.

@maxking was there any progress made on fixing the folding of other identification fields elsewhere by any chance? Or is this still an open issue? Thanks!

**Pull request merged**
